### PR TITLE
fix #7462: Guest window goes beyond the map edge on a spiral slide.

### LINKED ISSUE #7462: The view of a peep in his window goes beyond the edge of the map when he's on a spiral slide.
**OS:** Windows 8
**Version:** e.g. 0.1.3
**Commit/Build:** 2512e49

If you have a window open to watch a particular peep, and that peep gets on a spiral slide, the view window will jump out beyond the edge of the map while that peep is on the slide, then jump back as soon as he's getting off it.

As far as I can tell, the spiral slide is the only ride that does this.

- [ ? ] Reproducible in RCT2 (vanilla)?
- [ No ] Specific to multiplayer?

**Steps to reproduce:**
1. Click on a peep in the queue line for a spiral slide.
2. Watch him through the small window while he rides the slide.

**Screenshots / Video:**

arid heights 2018-04-28 21-55-08
arid heights 2018-04-28 21-56-01

### PR DESCRIPTION
**Pull request by moTechPlz**
Function uses both data structs to write and read so it cannot be a union.

### CODE REVIEW

**Comment by duncanspumpkin** [comment]
This does not look right. The data structs were meant to overlap and overwrite the memory hence why it was a union. I'm unsure if this change would be bug free.

**Comment by duncanspumpkin** [comment]
Argh please do not close PR's when making changes to them. I didn't see there was already conversation in the closed PR. #7741 #7740

**Comment by duncanspumpkin** [comment]
Okay i've found out what the real bug is its in the reading of the values. You should never read `coordinate_focus.y` without first masking the value with `VIEWPORT_FOCUS_Y_MASK`

**Comment by duncanspumpkin** [comment]
@moTechPlz https://github.com/duncanspumpkin/OpenRCT2/commit/360715d96a2e9bdc097ab8c1693df78e64d78de2 is how to prevent the bug from occuring.

File: src/openrct2-ui/windows/Guest.cpp [hunk 1]
```diff
@@ -759,10 +757,10 @@ void window_guest_viewport_init(rct_window* w){
 
     window_event_invalidate_call(w);
 
-    w->viewport_focus_coordinates.x = focus.coordinate.x;
-    w->viewport_focus_coordinates.y = focus.coordinate.y;
-    w->viewport_focus_coordinates.z = focus.coordinate.z;
-    w->viewport_focus_coordinates.rotation = focus.coordinate.rotation;
+    w->viewport_focus_coordinates.x = coordinate.x;
+    w->viewport_focus_coordinates.y = coordinate.y;
```
Comment (duncanspumpkin) [inline src/openrct2-ui/windows/Guest.cpp]: This would not pass the sprite_type information which is also stored in this memory address.

**Comment by moTechPlz** [comment]
> Argh please do not close PR's when making changes to them. I didn't see there was already conversation in the closed PR. #7741 #7740

I'm so sorry, i'm having trouble with github desktop, i think i need to learn the console.

> Okay i've found out what the real bug is its in the reading of the values. You should never read coordinate_focus.y without first masking the value with VIEWPORT_FOCUS_Y_MASK. duncanspumpkin/OpenRCT2@360715d is how to prevent the bug from occuring.

Ah, yes. That is how the code is suppose to function. Thank you.

**Comment by moTechPlz** [comment]
//todo:
- handle ride window viewport resize
- further testing

**Comment by moTechPlz** [comment]
Ok, this will fix the bug. Please tell me if there are things in need to do.

File: src/openrct2/interface/Window.h [hunk 2]
```diff
@@ -137,6 +137,12 @@ struct sprite_focus {
     uint8_t zoom; //0x489
 };
 
+union viewport_focus
+{
```
Comment (duncanspumpkin) [inline src/openrct2/interface/Window.h]: I'm not convinced that this is any better to the previous this all gets stuffed into the var_480 union of the window. This is just a subset of that union.

**Comment by AaronVanGeffen** [comment]
Thanks for the patch. I've taken the liberty to rebase the PR and apply `clang-format`. The fix you ended up with works as advertised, so I'll happily merge this.

**Pull request merged**